backends/aerospike.go

```diff
@@ -13,59 +15,111 @@ import (
+	if err != nil {
+		return nil, formatAerospikeError(err, "NEW_KEY")
+	}
+	return asKey, nil
+}
+
+type AerospikeBackend struct {
 	cfg     config.Aerospike
-	client  *as.Client
+	client  AerospikeDBClient
 	metrics *metrics.Metrics
 }
 
-func NewAerospikeBackend(cfg config.Aerospike, metrics *metrics.Metrics) *Aerospike {
+func NewAerospikeBackend(cfg config.Aerospike, metrics *metrics.Metrics) *AerospikeBackend {
+	if cfg.Host == "" {
+		log.Fatalf("Cannot connect to empty Aerospike host")
+	}
+	if cfg.Port <= 0 {
+		log.Fatalf("Cannot connect to Aerospike host at port %d", cfg.Port)
+	}
 	client, err := as.NewClient(cfg.Host, cfg.Port)
 	if err != nil {
-		log.Fatalf("Error creating Aerospike backend: %v", err)
-		panic("Aerospike failure. This shouldn't happen.")
+		log.Fatalf("Error creating Aerospike backend: %v", formatAerospikeError(err, "NewAerospikeBackend"))
+		panic("AerospikeBackend failure. This shouldn't happen.")
 	}
 	log.Infof("Connected to Aerospike at %s:%d", cfg.Host, cfg.Port)
 
-	return &Aerospike{
+	return &AerospikeBackend{
 		cfg:     cfg,
-		client:  client,
+		client:  &Aerospike{client},
 		metrics: metrics,
 	}
 }
 
-func (a *Aerospike) Get(ctx context.Context, key string) (string, error) {
-	asKey, err := as.NewKey(a.cfg.Namespace, setName, key)
+func (a *AerospikeBackend) Get(ctx context.Context, key string) (string, error) {
+	asKey, err := a.client.NewUuidKey(a.cfg.Namespace, key)
 	if err != nil {
 		return "", err
 	}
-	rec, err := a.client.Get(nil, asKey, "value")
+	rec, err := a.client.Get(asKey)
 	if err != nil {
 		return "", err
 	}
 	if rec == nil {
-		return "", errors.New("client.Get returned a nil record. Is aerospike configured properly?")
+		return "", errors.New("Aerospike GET. Nil record")
 	}
 	a.metrics.RecordExtraTTLSeconds(float64(rec.Expiration))
-	return rec.Bins[binValue].(string), nil
+
+	value, found := rec.Bins[binValue]
+	if !found {
+		return "", errors.New("Aerospike GET. No 'value' bucket found")
+	}
+
+	str, isString := value.(string)
+	if !isString {
+		return "", errors.New("Aerospike GET. Retrieved value is not a string")
+	}
+
+	return str, nil
 }
 
-func (a *Aerospike) Put(ctx context.Context, key string, value string, ttlSeconds int) error {
-	asKey, err := as.NewKey(a.cfg.Namespace, setName, key)
+func (a *AerospikeBackend) Put(ctx context.Context, key string, value string, ttlSeconds int) error {
+	asKey, err := a.client.NewUuidKey(a.cfg.Namespace, setName)
 	if err != nil {
 		return err
 	}
 	if ttlSeconds == 0 {
 		ttlSeconds = a.cfg.DefaultTTL
 	}
-	bins := as.BinMap{
-		binValue: value,
-	}
-	err = a.client.Put(&as.WritePolicy{
-		Expiration: uint32(ttlSeconds),
-	}, asKey, bins)
+	return a.client.Put(asKey, value, ttlSeconds)
+}
+
+func formatAerospikeError(err error, caller string) error {
 	if err != nil {
-		return err
+		if aerr, ok := err.(ase.AerospikeError); ok {
+			return fmt.Errorf("%s Aerospike error: %s. Code: %d", caller, aerr.Error(), aerr.ResultCode())
```

Comment:
Is `caller` really needed? The caller/method name that returned the error wouldn't make any sense to the user IMHO. I'd assume that the error message itself should be indicative of where the problem is. Same for the ResultCode. I believe, returning just the error message as is should suffice.

The Aerospike Go library has an error message associated with most of the Result Codes (https://github.com/aerospike/aerospike-client-go/blob/d3734af6b6fc13aa6e05c1471a1e93dc90743277/types/error.go#L77-L83) and for the ones where it doesn't have an error message associated, it does return the ResultCode as part of the message (https://github.com/aerospike/aerospike-client-go/blob/d3734af6b6fc13aa6e05c1471a1e93dc90743277/types/error.go#L67).

Reply:
I believe for our investigation is better to explicitly specify the method but you raise a good point. Let's discuss.